servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java

```diff
@@ -240,7 +236,9 @@ private DefaultSingleAddressHttpClientBuilder(@Nullable final U address,
 
         U address() {
             assert builder.address != null : "Attempted to buildStreaming with an unknown address";
-            return proxyAddress != null ? proxyAddress : builder.address;
+            return proxyAddress != null ? proxyAddress :
+                    // the builder can be modified post-context creation, therefore proxy can be set
+                    (builder.proxyAddress != null ? builder.proxyAddress : builder.address);
```

Comment:
Referencing `builder.proxyAddress` is unexpected. The motivation of `HttpClientBuildContext` is to capture mutable the state of the builder, so it can be re-used in subsequent `build()` without state unexpectedly changing. Note that `builder.address` is `final` so it will not change, and IIUC that is why it is referenced here.

Reply:
From my observations it looks like the assumption here is that the `builder` is **already a copy** owned by the context ([source](https://github.com/apple/servicetalk/blob/main/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java#L412)). The codebase uses that fact with the concept of `MultiAddressHttpClientBuilder.SingleAddressInitializer` where the cloned builder is saved during the `ClientFactory` creation ([source](https://github.com/apple/servicetalk/blob/main/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java#L96)) and later modified for each individually built client ([source](https://github.com/apple/servicetalk/blob/main/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java#L225-L244)).

Finally, those changes on the copied builder which is embedded as part of the captured `HttpClientBuildContext` (and may, as encouraged by current design, be later modified) are used to build the actual underlying client ([source](https://github.com/apple/servicetalk/blob/main/servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java#L271-L354)).

Perhaps the design isn't ideal due to the entangled nature of the context:
- immutable: created from a mutable builder using a copy method,
- contains a mutable copied builder that is expected to be modified,
- needs to be reconciled during build time with the mutable state in order to achieve the flexibility required by the multi/partitioned builders.

This issue could be subject of a different PR, but all in all it feels that the end goal of preventing concurrent access is taken care of.

Reply:
> preventing concurrent access 

It looks like our current usage is safe, but do we need to be more clear throughout the ST API about which instances can be shared between threads and which are non-threadsafe? Are users likely to be unsure which objects they can share? Builders are typically not thread safe as they are not generally meant to be shared. The "copy-on-write" solution used here for a partially complete builder seems OK.

Reply:
I would love the APIs to expose something like https://jcip.net/annotations/doc/net/jcip/annotations/NotThreadSafe.html, but I think it's out of scope of this PR.

With regards to our case: as long as ` io.servicetalk.client.api.DefaultClientGroup#get` is thread safe, we do not have an issue. Users also need to make sure they don't reuse the builder from multiple threads, but being super explicit about it in the documentation would be beneficial in my opinion.

Reply:
Correct, I added a test for it.